odl/space/space_utils.py

```diff
@@ -281,6 +284,303 @@ def rn(shape, dtype=None, impl='numpy', **kwargs):
     return rn
 
 
+class auto_weighting(OptionalArgDecorator):
+
+    """Make an unweighted adjoint automatically account for weightings.
+
+    This decorator can be used in `Operator` subclasses to automatically
+    perform weighting of the adjoint operator based on the space
+    weightings. It is used as ::
+
+        class MyOperator(odl.Operator):
+
+            @auto_weighting
+            @property
+            def adjoint(self):
+                return MyOperatorUnweightedAdjoint(...)
+
+    In this context, the term "unweighted adjoint" of an operator ``A``
+    refers to the the adjoint of ``A`` defined between **unweighted
+    Euclidean spaces**. In particular, discretized ``L^p`` type spaces are
+    interpreted as weighted variants of ``R^n`` or ``C^n`` spaces.
+
+    For instance, an (m x n) real matrix ``M`` is also an operator
+    ``M: R^n -> R^m``, and its unweighted adjoint is defined by the
+    transposed matrix ``M^T: R^m -> R^n``.
+
+    Similarly, when an integration operator ``A: L^2(I) -> R`` on an
+    interval ``I`` is discretized using ``n`` grid points, it can be seen
+    as a summation operator ``S: R^n -> R, S(x) = dx * sum(x)`` scaled
+    by the subinterval size ``dx``.
+    Here, the unweighted adjoint is the adjoint ``S^*: R -> R^n`` of ``S``,
+    given by ``S^*(c) = dx * (c, ..., c)``.
+    The weighting to correctly define the *weighted* adjoint
+    ``A^*: R -> L^2(I)`` is taken care of by this decorator.
+
+    Parameters
+    ----------
+    unweighted_adjoint : `Operator`
```

Comment:
Shouldn't the type of this be "method" or something like that?

Reply:
Technically yes. I'll change it.